File: DeezerSample/DeezerSample/DeezerObjectListViewController.m
//
//  DeezerObjectListViewController.m
//  DeezerSample
//
//  Created by GFaure on 09/05/2014.
//  Copyright (c) 2014 Deezer. All rights reserved.
//

#import "DeezerObjectListViewController.h"
#import "DeezerItemViewController.h"
#import "DZRModel.h"
#import "DZRRequestManager.h"
#import "DZRCancelable.h"
#import "DeezerAudioPlayerController.h"

@interface DeezerObjectListViewController () <UITableViewDataSource, UITableViewDelegate>
{
    IBOutlet UITableView *objectsView;
}

@property(nonatomic, strong) DZRRequestManager *manager;
@property(nonatomic, strong) DZRRequestManager *illustrationManager;
@property(nonatomic, strong) DZRObjectList *list;
@property(nonatomic, strong) id<DZRPlayable> playable;
@property(nonatomic, strong) NSArray *data;
@property(nonatomic, strong) NSCache *imageCache;
@end

@implementation DeezerObjectListViewController

- (id)initWithDZRObjectList:(DZRObjectList *)list
{
    return [self initWithDZRObjectList:list fromPlayable:nil];
}

- (id)initWithDZRObjectList:(DZRObjectList *)list fromPlayable:(id)playable
{
    NSString *nibName;
    if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPhone) {
        nibName = @"DeezerObjectListViewController_iPhone";
    } else {
        nibName = @"DeezerObjectListViewController_iPad";
    }

    if (self = [super initWithNibName:nibName bundle:nil]) {
        self.imageCache = [NSCache new];
        self.imageCache.countLimit = 50;
        self.manager = [[DZRRequestManager defaultManager] subManager];
        self.list = list;
        self.playable = playable;
    }
    return self;
}

- (void)dealloc
{
    self.data = nil;
    [self.manager cancel];
    self.manager = nil;
    self.imageCache = nil;
    self.list = nil;
}

- (void)setList:(DZRObjectList *)list
{
    if (_list != list) {
        _list = list;
        [_list allObjectsWithManager:self.manager callback:^(NSArray *objs, NSError *error) {
            self.data = objs;
            [objectsView reloadData];
            [self scrollViewDidEndDecelerating:objectsView];
        }];
    }
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [objectsView reloadData];
    [self scrollViewDidEndDecelerating:objectsView];
}

#pragma mark UITableVieWDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.data.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *reuseIdentifier = @"item";
    DZRObject *o = [self.data objectAtIndex:indexPath.row];

    UITableViewCell *cell = [objectsView dequeueReusableCellWithIdentifier:reuseIdentifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:reuseIdentifier];
    }

    cell.textLabel.text = [o description];
    cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    cell.imageView.image = nil;

    return cell;
}

#pragma mark UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    DZRObject *o = [self.data objectAtIndex:indexPath.row];

    if (self.delegate && [self.delegate objectListViewControllor:self didSelectObjects:@[o]]) {
        [self.navigationController popViewControllerAnimated:YES];
    }
    else {
        if (self.playable != nil) {
            DeezerAudioPlayerController *playerVC = [[DeezerAudioPlayerController alloc] initWithPlayable:self.playable startIndex:indexPath.row];
            [self.navigationController pushViewController:playerVC animated:YES];
        }
        else {
            DeezerItemViewController *details = [[DeezerItemViewController alloc] initWithDZRObject:o];
            [self.navigationController pushViewController:details animated:YES];
        }
    }
}

#pragma mark UIScrollViewDelegate

- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView
{
    self.illustrationManager = [self.manager subManager];
    [[objectsView visibleCells] enumerateObjectsUsingBlock:^(UITableViewCell *cell, NSUInteger idx, BOOL *stop) {
        DZRObject *o = [self.data objectAtIndex:[objectsView indexPathForCell:cell].row];
        if (o.isIllustratable) {
            UIImage *cachedImage = [self.imageCache objectForKey:o];
            __weak NSCache *cache = self.imageCache;
            if (!cachedImage) {
                [(DZRObject <DZRIllustratable> *) o
                        illustrationWithRequestManager:self.manager
                                              callback:^(UIImage *illustration, NSError *error)
                        {
                            if (illustration) {
                                [cache setObject:illustration forKey:o];
                                cell.imageView.image = illustration;
                                [cell setNeedsLayout];
                            }
                        }];
            }
            else {
                cell.imageView.image = cachedImage;
                [cell setNeedsLayout];
            }
        }
    }];
}

- (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
{
    [self.illustrationManager cancel];
    self.illustrationManager = nil;
}
@end

File: DeezerSampleSwift/DeezerSampleSwift/ViewControllers/DeezerObjectListViewController.swift
import UIKit

class DeezerObjectListViewController: UIViewController {
    
    @IBOutlet private weak var tableView: UITableView!
    
    private var data: [Any]? {
        didSet {
            tableViewDelegate?.data = data
        }
    }
    private var deezerObjectList: DZRObjectList? {
        didSet {
            getData()
        }
    }
    
    private var tableViewDelegate: DeezerObjectTableViewDelegate?
    var object: DeezerObject?

    override func viewDidLoad() {
        super.viewDidLoad()
        title = object?.title
        setupTableView()
        getObjectList()
    }
    
    func getObjectList() {
        guard let type = self.object?.type else {
            return
        }
        
        
        type.getObjectList(object: object?.object, callback: { [weak self] (deezerObjectList, error) in
            guard let deezerObjectList = deezerObjectList, let strongSelf = self else {
                print(error.debugDescription )
                return
            }
            strongSelf.deezerObjectList = deezerObjectList
        })
    }
    
    func getData() {
        guard let objectList = self.deezerObjectList else {
            return
        }
        DeezerManager.sharedInstance.getData(fromObjectList: objectList) {[weak self] (data, error) in
            guard let data = data, let strongSelf = self else {
                if let error = error {
                    self?.present(error: error)
                }
                return
            }
            
            strongSelf.data = data
            strongSelf.tableView.reloadData()
        }
    }
    
    func setupTableView() {
        tableViewDelegate = DeezerObjectTableViewDelegate(viewController: self, object: object)
        tableView.delegate = tableViewDelegate
        tableView.dataSource = self
        tableView.tableFooterView = UIView()
        tableView.rowHeight = 50
    }
}

// MARK: - UITableViewDataSource

extension DeezerObjectListViewController: UITableViewDataSource {
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return data?.count ?? 0
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = UITableViewCell()

        guard let deezerObject = data?[indexPath.row] as? DZRObject else {
            return cell
        }
        
        cell.textLabel?.text = deezerObject.description
        
        return cell
    }
    
}
